contents/staticAnalysis/PMD.md

```diff
@@ -0,0 +1,86 @@
+
+#### Suppress Warnings
+PMD supports suppressing warnings in four ways:
+
+- Annotations
+- Comments
+- Violation Suppress Regex
+- Violation Suppress XPath
+
+The details are described [here](https://pmd.github.io/pmd-5.5.3/usage/suppressing.html).
+
+#### Command Line Usage
+PMD can be launched by using command line with various arguments. For details, please refer to [Running PMD via command line](https://pmd.github.io/pmd-5.5.3/usage/running.html).
+
+#### Integration with Build Automation Tools
+- [Ant task usage](https://pmd.github.io/pmd-5.5.3/usage/ant-task.html)
+- [The PMD Plugin in Gradle](https://docs.gradle.org/current/userguide/pmd_plugin.html)
+- [Maven 1 PMD plugin](https://pmd.github.io/pmd-5.5.3/usage/maven-plugin.html)
+- [Maven 2 PMD plugin](https://pmd.github.io/pmd-5.5.3/usage/mvn-plugin.html)
+
+#### Integration with IDEs
+PMD can be integrated with most of IDEs, inlcuding BlueJ, CodeGuide, Eclipse, eclipse-pmd, Emacs, Gel, IntelliJ IDEA, IntelliJ IDEA - QAPlug, JBuilder, JCreator, JDeveloper, JEdit, Maven, Maven 2, NetBeans, TextPad, WebLOgic Workshop 8.1.x. For instruction to integrate with those IDEs, please refer to [PMD integrations](https://pmd.github.io/pmd-5.5.3/usage/integrations.html).
+
+### Available Rulesets
+There is not pre-defined rules set. You need to [define your rule sets](https://pmd.github.io/pmd-5.5.3/customizing/howtomakearuleset.html) by yourself.
+
+PMD has organised rules into different categories. For example, the rules for Java has been categorised into 26 sections, which will help you quickly find the rules that you want.
+
+You may refer to the [Features](#features) section to view the rules according to your languages and refer to [Configuration](#configuration) section to configure your rulesets.
+
+### How does it work
+PMD use [JavaCC](http://javacc.org) to parse your code to a [AST(Abstract Syntax Tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) and visited it recursively ([more details](https://pmd.github.io/pmd-5.4.1/customizing/howitworks.html)). Thus, one of the requirements for PMD to work is that the code must be in valid Java syntax. You can view the syntax tree by using [bin/designer.bat](https://pmd.github.io/pmd-5.5.3/customizing/howtowritearule.html)
+
+### Developer (Customisation)
+- [How to write a PMD rule](https://pmd.github.io/pmd-5.5.3/customizing/howtowritearule.html) (A quicker way to write rule sets [Using XPath in PMD](https://pmd.github.io/pmd-5.4.1/customizing/xpathruletutorial.html), [XPath Rule tutorial](https://pmd.github.io/pmd-5.4.1/customizing/xpathruletutorial.html))
+- [Add supports for new language](https://pmd.github.io/pmd-5.4.1/customizing/new-language.html) ([Add supports for CPD](https://pmd.github.io/pmd-5.4.1/customizing/cpd-parser-howto.html))
+
+### Reference
+- [PMD](https://pmd.github.io/)
+- [PMD Github](https://github.com/pmd/pmd)
+- [StackOverflow PMD](https://stackoverflow.com/questions/tagged/pmd)
```

Comment:
Technically, _references_ should be referenced explicitly in the main text. For example, something like this:

> The tool X is currently the most popular tool in the marked [source: [X Dev Guide]()]. However, tool Y is also ...
> ...
> ### References
> * [X Dev Guide] [_A Guide to Developing on X_]() - This is the official Developer Guide for X

In these docs, it is better to have a **Resources** section instead. e.g.

> ### Resources
> * [_Introduction to X_]() - A beginner friendly tutorial written by X dev team. If you are totally new to X, this is the best place to start.
> *